### ISSUE TITLE
HTTPSE is the heaviest operation on page load

### ISSUE THREAD

**Issue by garvankeeley** [brave/browser-ios#310]
Other than UIWebView internals, HTTPSE sql db lookup is our heaviest operation. 
I ran a quick profile and found HTTPSE db lookup is clearly top-of-the-heap. 
TODO for me: put in a good capture from the profiler, I think if I load an http site with lots of subresource loading (requiring lots of Https-E candidate lookups) I can demonstrate a 1sec added page load time.

**Comment by garvankeeley** [comment]
Added a fix for repeated loads here, HTTPSE is no-cost when hitting sites seen already in the current session: 
https://github.com/brave/browser-ios/pull/335

**Comment by garvankeeley** [comment]
@diracdeltas I still haven't got a nice profile for this, I will do, but heads up that hitting the SQL db on first load of a site is adding 15-20% to load time. (in-mem caching of HTTPSE lookups in iOS app ensures that *repeating* page loads are pretty much never hitting the DB)

Not sure if the SQL queries themselves could be optimized, or if we can have some bot that browses the Alexa top 1000, and creates a bloom filter from that, and we can check the in-mem bloom filter before hitting the DB.

**Comment by garvankeeley** [comment]
@alexwykoff this is an interesting one to watch out for. 

In the next beta, I will have https://github.com/brave/browser-ios/pull/335 landed, at that point, the bug will only be visible the first time you visit a page (until app restart). On fresh app start, when you load HTTP (non HTTPS) sites with lots of resources (pages like cnn.com make ~200 requests during load), each one of those requests gets checked in the HTTPSE database for upgrading from http to https. The SQL DB for this is >10MB and the queries are fuzzy comparisons IIRC (not simple 1 to 1 key lookups), and on mobile this adds up to a performance headache.

**Comment by garvankeeley** [comment]
Another option is to regress this feature to only try HTTPSE lookup on the top-level doc. But I am rather attached to the coolness of HTTPSE upgrades of the resource loads.

**Comment by diracdeltas** [comment]
the HTTPS Everywhere rules are also available as JSON on AWS but i'm guessing that won't be any faster. we added a LRU cache in desktop-browser for recently-used rulesets, maybe that would help here?

**Comment by garvankeeley** [comment]
We have in-memory caches, so the hit is when visiting a site for the first time in an app session.

**Comment by garvankeeley** [comment]
@SergeyZhukovsky I switched to the cpp code, on iPad3 sqlite3_profile() reports these timings on the 1st query, in milliseconds. Here is the first few:

select ids from targets where host like 'thestar.com' ^^342 
select ids from targets where host like 'thestar.com' ^^341 
select ids from targets where host like '*.thestar.com' or host like 'www.thestar.com' ^^449 
select ids from targets where host like 'apple.com' ^^299 
select ids from targets where host like 'xkcd.com' ^^298 
select ids from targets where host like '*.googleadservices.com' or host like 'www.googleadservices.com' ^^451

**Comment by SergeyZhukovsky** [comment]
@garvankeeley how it was before cpp? Is it much faster now?

**Comment by garvankeeley** [comment]
A little bit faster, because of the aggregated 'or' query in the first step, we can assume on this device >300ms per query, and we see a 2-part query on the 3rd line above is 450ms.

**Comment by SergeyZhukovsky** [comment]
Good, in any case it is better to do in one request, what times with caches?

**Comment by garvankeeley** [comment]
HTTPSE requires synchronous DB calls, which regardless of the optimizations thrown at the DB will result in a performance bottleneck.

**Comment by garvankeeley** [comment]
LIKE operators are expensive in SQLite as it bypasses index, not totally sure why these are needed, it looks like a string match operation to me.

UNIQUE fields are auto-indexed in sqlite, so if the HTTPS-E file is generated with this we will get indexed query results.

Changing the SQL query to equality instead of like (which has minimal effect on its own), and then doing:
```
 sqlite3_exec(httpse_db_, "CREATE INDEX IF NOT EXISTS hostidx ON targets(host)", 0, 0, &err)
```

Reduces average query time for 250ms on iPhone 5s to <1ms

This has zero effect, but httpse is currently sync on iOS:
```
 sqlite3_exec(httpse_db_, "PRAGMA synchronous=OFF", 0, 0, &err) != SQLITE_OK)
```

**Comment by garvankeeley** [comment]
TLDR: 
1) make targets.host a UNIQUE field when creating HTTPSE db,
2) don't use like
 and 250 ms queries go to <1 ms on my test device.

**Comment by bbondy** [comment]
@SergeyZhukovsky

**Comment by garvankeeley** [comment]
On my iPhone 5c, cnn.com is taking ~8s to load with HTTPSE off, and ~17s to load with HTTPSE on and non-indexed, and ~12s to load with indexing on. 

@diracdeltas  sorry, my DB index fix may not be the last tweak -not the smoking gun i hoped, so although it will improve things, more smoke comes from over yonder. There is a tiny chance I might find another DB tweak. I seriously doubt it though, the DB is very basic.
/me more profiling...

**Comment by garvankeeley** [comment]
Profiling seems to confirm 200-300ms spent in first DB lookup code, and 300ms spent in second lookup (that is including the JSON parsing). So far I can account for 600ms of added load time.

**Comment by garvankeeley** [comment]
When running in instruments profiler, I can see 800-1000 ms of a ~10 sec load time on cnn.com is HTTPSE sqlite calls, there is no other HTTPSE code (JSON parsing, etc.) that is significant in the profile.
The second table already has a primary key set on 'id' column, which is auto indexed in sqlite, so that table lookup is optimal.
I see occasional sql query spikes when CPU and disk I/O is high, queries that are normally 1ms, jump to 100ms. On a heavy site like cnn.com, the device is at maximal load.

For interest, I tried loading the 'targets' table as a hash table in memory, and RAM spiked 10-15 MB (didn't look too closely, simply too much to be acceptable)

**Comment by garvankeeley** [comment]
With the latest code optimizations (commits to follow), we see HTTPSE takes no longer than Adblock+TP (testing cnn.com on iPhone 5c, load time ~10 sec):

~800 ms HTTPSE:
<img width="652" alt="screenshot 2016-08-16 23 39 33" src="https://cloud.githubusercontent.com/assets/5495083/17723803/dabee99e-640a-11e6-9d6d-3019e6c477ba.png">

~200 ms TP:
<img width="694" alt="screenshot 2016-08-16 23 39 21" src="https://cloud.githubusercontent.com/assets/5495083/17723805/dd60c9c4-640a-11e6-86fe-2a5706ed1e1d.png">

~800 ms Adblock:
<img width="673" alt="screenshot 2016-08-16 23 39 06" src="https://cloud.githubusercontent.com/assets/5495083/17723807/df6a6536-640a-11e6-8450-c51f7775bd91.png">

**Comment by garvankeeley** [comment]
Pinging @diracdeltas @SergeyZhukovsky @bbondy to see last comment on this. 
I am moving on to cleaning my changes, getting them in as swift first (because they are minor tweaks to the existing code), then to the c++ branch I am working on, so code can be shared easily in future (the C++ code will land after, and we can make sure there is no regression).

**Comment by garvankeeley** [comment]
Marking this fixed by: https://github.com/brave/browser-ios/commit/77211eb7493ad29d53fe179a4d9e2238b96c772b
